gusto/time_discretisation/explicit_runge_kutta.py

```diff
@@ -114,45 +143,85 @@ def setup(self, equation, apply_bcs=True, *active_labels):
         """
         super().setup(equation, apply_bcs, *active_labels)
 
-        if not self.increment_form:
-            self.field_i = [Function(self.fs) for i in range(self.nStages+1)]
+        if self.rk_formulation == RungeKuttaFormulation.predictor:
+            self.field_i = [Function(self.fs) for _ in range(self.nStages+1)]
+        elif self.rk_formulation == RungeKuttaFormulation.increment:
+            self.k = [Function(self.fs) for _ in range(self.nStages)]
+        elif self.rk_formulation == RungeKuttaFormulation.linear:
+            self.field_rhs = Function(self.fs)
         else:
-            self.k = [Function(self.fs) for i in range(self.nStages)]
+            raise NotImplementedError(
+                'Runge-Kutta formulation is not implemented'
+            )
 
     @cached_property
     def solver(self):
-        if self.increment_form:
+        if self.rk_formulation == RungeKuttaFormulation.increment:
             return super().solver
-        else:
+
+        elif self.rk_formulation == RungeKuttaFormulation.predictor:
             # In this case, don't set snes_type to ksp only, as we do want the
-            # outer Newton iteration
+            # outer Newton iteration. This is achieved by not calling the
+            # "super" method, in which the default snes_type is set to ksp_only
             solver_list = []
 
             for stage in range(self.nStages):
                 # setup linear solver using lhs and rhs defined in derived class
                 problem = NonlinearVariationalProblem(
                     self.lhs[stage].form - self.rhs[stage].form,
-                    self.field_i[stage+1], bcs=self.bcs)
+                    self.field_i[stage+1], bcs=self.bcs
+                )
                 solver_name = self.field_name+self.__class__.__name__+str(stage)
                 solver = NonlinearVariationalSolver(
                     problem, solver_parameters=self.solver_parameters,
-                    options_prefix=solver_name)
+                    options_prefix=solver_name
+                )
                 solver_list.append(solver)
             return solver_list
 
+        elif self.rk_formulation == RungeKuttaFormulation.linear:
+            # In this case, don't set snes_type to ksp only, as we do want the
```

Comment:
Same as above

Reply:
Probably a very good point. I think we've decided to resolve this in #560